src/transformers/tokenization_utils_fast.py

```diff
@@ -256,7 +256,10 @@ def convert_tokens_to_ids(self, tokens: Union[str, List[str]]) -> Union[int, Lis
 
     def _convert_token_to_id_with_added_voc(self, token: str) -> int:
         index = self._tokenizer.token_to_id(token)
-        if index is None:
+        # TODO: it looks like the '' token is not re-added when retraining
+        # the tokenizer in tests, and we fall into an infinite reursion
+        # trying to convert unknown token to id
+        if index is None and token != self.unk_token:
```

Comment:
I think the issue comes down to this:

```python
import string
from transformers import BertTokenizerFast

tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased")
tokenizer._tokenizer.add_special_tokens([''])
print(tokenizer._tokenizer.token_to_id(''))
```

Adding any other token (e.g. `<|test|>`) works fine, but the empty string doesn't work as a token.

Reply:
Ok, the actual discrepancy is between how slow and fast tokenizers handle adding the `''` token:

**Slow**

```python
import string
from transformers import BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-cased")
tokenizer.add_tokens([''])
print(tokenizer.convert_tokens_to_ids(''))
#=> 28996
```

**Fast**

```python
import string
from transformers import BertTokenizerFast

tokenizer = BertTokenizerFast.from_pretrained("bert-base-cased")
tokenizer.add_tokens([''])
print(tokenizer.convert_tokens_to_ids(''))
#=> 100 (unknown)
```

Reply:
Yep exactly. Which is why in #21250 I set any unknown token to `""`. This was such a headache.
In the `whisper-large` version, I added `""` to the list of world in the vocab, and set `unk_token = ""` which gave the correct ID, but it is a bit confusing.

Reply:
The fast version should, by default already do this. If you try with `whisper-tiny.en` you should have a correct output. 
Also when you use `GPT2TokenizerFast` you should also have the correct output (OOV are decoded as `""`) no

Reply:
It works fine with `whisper-tiny.en`, but breaks for `whisper-tiny`, which configures `""` as the unknown token.

To be clear, the issue is specifically in tests where we retrain the tokenizer:

```python
import string
from transformers import WhisperTokenizerFast

tokenizer = WhisperTokenizerFast.from_pretrained("openai/whisper-tiny")
vocabulary = string.ascii_letters + string.digits + " "
tokenizer = tokenizer.train_new_from_iterator(vocabulary, vocab_size=len(vocabulary), show_progress=False)
tokenizer.all_special_ids
```

When we remove the temporary check here:

https://github.com/huggingface/transformers/blob/53cbd1b35ecc85356067b6757cda6ec24e22fcb6/src/transformers/tokenization_utils_fast.py#L259-L262

we get an infinite recursion, that's because the `''` token is no longer in the tokenizer.

Reply:
If you want to reproduce locally, remove that check and run `pytest tests/pipelines/test_pipelines_automatic_speech_recognition.py -k 'test_pt_WhisperConfig_WhisperForConditionalGeneration_WhisperTokenizer_WhisperFeatureExtractor'` :)

Reply:
Yep, I think we should remove the unk_token (the way it is defined) from the multilingual models. They should behave the same way as `whisper-tiny.en`

Reply:
No it should't, we will leave `''` in the vocab but just set `unk_token = <|endoftext|>`

Reply:
Ah, it is in the vocab too, not just the special tokens, perfect!